## 1) Initialisation & Chargement des données

In [2]:
import pyspark
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, lit, datediff, max as spark_max, sum as spark_sum, countDistinct

In [3]:
spark = SparkSession.builder \
    .appName("Spark_Final_Project") \
    .master("local[*]") \
    .getOrCreate()

26/02/10 23:04:21 WARN Utils: Your hostname, MacBook-Pro.local resolves to a loopback address: 127.0.0.1; using 192.168.0.29 instead (on interface en0)
26/02/10 23:04:21 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
26/02/10 23:04:21 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [4]:
df_raw = spark.read \
    .option("header", True) \
    .option("inferSchema", True) \
    .option("delimiter", ";") \
    .csv("../data/raw/Online_Retail_CSV.csv")

df_raw.printSchema()
df_raw.show(5)

root
 |-- InvoiceNo: string (nullable = true)
 |-- StockCode: string (nullable = true)
 |-- Description: string (nullable = true)
 |-- Quantity: integer (nullable = true)
 |-- InvoiceDate: string (nullable = true)
 |-- UnitPrice: string (nullable = true)
 |-- CustomerID: integer (nullable = true)
 |-- Country: string (nullable = true)

+---------+---------+--------------------+--------+----------------+---------+----------+--------------+
|InvoiceNo|StockCode|         Description|Quantity|     InvoiceDate|UnitPrice|CustomerID|       Country|
+---------+---------+--------------------+--------+----------------+---------+----------+--------------+
|   536365|   85123A|WHITE HANGING HEA...|       6|01/12/2010 08:26|     2,55|     17850|United Kingdom|
|   536365|    71053| WHITE METAL LANTERN|       6|01/12/2010 08:26|     3,39|     17850|United Kingdom|
|   536365|   84406B|CREAM CUPID HEART...|       8|01/12/2010 08:26|     2,75|     17850|United Kingdom|
|   536365|   84029G|KNITTED UNI

In [5]:
print(f"Il y a {df_raw.count()} lignes.")
print(f"Il y a {len(df_raw.columns)} colonnes.")

Il y a 541909 lignes.
Il y a 8 colonnes.


## 2) Exploration & Prétraitement

## A) Analyse Descriptive

Tout d'abord, regardons combien il y a de clients référencés dans cette base de données :

In [6]:
print(f"Il y a {df_raw.select('CustomerID').distinct().count()} clients.")

Il y a 4373 clients.


Maintenant, regardons le nombre total de transactions :

In [7]:
transaction_nb = df_raw.select('InvoiceNo').dropDuplicates().count()
print(f"Il y a {transaction_nb} transactions uniques.")

Il y a 25900 transactions uniques.


Regardons de plus près la distribution de la variable Quantity :

In [8]:
df_raw.select("Quantity").describe().show()

+-------+------------------+
|summary|          Quantity|
+-------+------------------+
|  count|            541909|
|   mean|  9.55224954743324|
| stddev|218.08115785023418|
|    min|            -80995|
|    max|             80995|
+-------+------------------+



Nous observons qu’en moyenne, une ligne de commande contient environ 9 produits.  
Cependant, l’écart-type est relativement élevé (218), ce qui indique une forte dispersion des valeurs et la présence d’outliers dans la variable Quantity.  
En effet, certaines transactions présentent des quantités extrêmement faibles ou élevées, allant jusqu’à –80 995 et +80 995 unités.  
Ces valeurs extrêmes contribuent fortement à l’augmentation de l’écart-type et sont susceptibles de correspondre à des retours de produits, des annulations de commandes ou des commandes en gros.

Regardons egalement la distribution de la variable UnitPrice :

In [9]:
df_raw.select("UnitPrice").describe().show()

+-------+------------------+
|summary|         UnitPrice|
+-------+------------------+
|  count|            541909|
|   mean|29.921163668665333|
| stddev| 595.7455525989117|
|    min|         -11062,06|
|    max|             99,96|
+-------+------------------+



Nous observons que le prix unitaire moyen d’un produit est d’environ 30 unités monétaires.  
Toutefois, l’écart-type est particulièrement élevé (596), ce qui traduit une forte dispersion des valeurs et la présence d’outliers dans la variable UnitPrice.  
En effet, certaines transactions présentent des valeurs négatives, pouvant atteindre –11 062, qui ne correspondent pas à des prix réels mais sont probablement liées à des remboursements, des annulations de commandes ou des écritures de correction comptable.  
Ces valeurs extrêmes contribuent fortement à l’augmentation de l’écart-type.

Maintenant, intéressons-nous au nombre de pays clients :

In [10]:
client_countries = df_raw.select("Country").dropDuplicates().count()
print(f"Il y a {client_countries} pays qui sont clients.")

Il y a 38 pays qui sont clients.


Regardons le nombre de commandes par pays : 

In [11]:
df_raw.select("InvoiceNo", "Country").dropDuplicates().groupby("Country").count().orderBy("count", ascending=False).show()

+---------------+-----+
|        Country|count|
+---------------+-----+
| United Kingdom|23494|
|        Germany|  603|
|         France|  461|
|           EIRE|  360|
|        Belgium|  119|
|          Spain|  105|
|    Netherlands|  101|
|    Switzerland|   74|
|       Portugal|   71|
|      Australia|   69|
|          Italy|   55|
|        Finland|   48|
|         Sweden|   46|
|         Norway|   40|
|Channel Islands|   33|
|          Japan|   28|
|         Poland|   24|
|        Denmark|   21|
|         Cyprus|   20|
|        Austria|   19|
+---------------+-----+
only showing top 20 rows



Nous observons que le jeu de données comprend des clients provenant de 38 pays différents, ce qui indique une base de clientèle internationale.  
Toutefois, une très grande proportion des transactions est réalisée au Royaume-Uni, avec plus de 22 000 factures uniques, tandis que le deuxième pays, l’Allemagne, ne compte que 603 transactions.  
Cela met en évidence un fort déséquilibre géographique, le Royaume-Uni dominant largement le jeu de données par rapport aux autres pays.

Regardons maintenant le nombre total de commandes par clients, cela nous permettra de distinguer les clients habituels des clients ponctuels : 

In [12]:
df_raw.select("CustomerID", "InvoiceNo").dropDuplicates().groupBy("CustomerID").count().orderBy("count", ascending=False).show()

+----------+-----+
|CustomerID|count|
+----------+-----+
|      NULL| 3710|
|     14911|  248|
|     12748|  224|
|     17841|  169|
|     14606|  128|
|     15311|  118|
|     13089|  118|
|     12971|   89|
|     14527|   86|
|     13408|   81|
|     14646|   77|
|     16029|   76|
|     16422|   75|
|     14156|   66|
|     13798|   63|
|     18102|   62|
|     13694|   60|
|     15061|   55|
|     17450|   55|
|     16013|   54|
+----------+-----+
only showing top 20 rows



On remarque la présence d’un groupe CustomerID = NULL avec un nombre élevé de commandes. Cette observation suggère la présence de transactions sans identification client, ce qui motive l’analyse des valeurs manquantes présentée dans la section suivante.

## B) Vérifications de valeurs manquantes

Avant toute analyse plus poussée, il est important de vérifier les valeurs nulles :  

In [13]:
from pyspark.sql.functions import col, sum, when

df_raw.select([sum(when(col(c).isNull(), 1 ).otherwise(0)).alias(c) for c in df_raw.columns]).show()

+---------+---------+-----------+--------+-----------+---------+----------+-------+
|InvoiceNo|StockCode|Description|Quantity|InvoiceDate|UnitPrice|CustomerID|Country|
+---------+---------+-----------+--------+-----------+---------+----------+-------+
|        0|        0|       1454|       0|          0|        0|    135080|      0|
+---------+---------+-----------+--------+-----------+---------+----------+-------+



Nous observons la présence de valeurs manquantes dans deux colonnes sur huit : Description et CustomerID.  
La colonne Description contient un nombre relativement limité de valeurs manquantes (1 454), ce qui reste marginal au regard de la taille du jeu de données et n’impactera pas significativement l’analyse, d’autant plus que cette variable ne sera pas utilisée directement dans les modèles.  
En revanche, la colonne CustomerID présente un nombre important de valeurs manquantes (135 080), ce qui constitue un enjeu majeur pour l’analyse, notamment pour la segmentation client. Ces lignes devront faire l’objet d’un traitement spécifique lors de la phase de nettoyage des données.

Pour pallier au problème de valeurs manquantes mais egalement de prix et de quantité négatives, nous allons créer un nouveau dataset filtré sur ces trois conditions :  
1 - Les valeurs de CustomerID doivent êtres Non Nulles  
2 - Les valeurs de Quantity doivent êtres > 0  
3 - Les valeurs unitaires des produits doivent êtres > 0  

Nous créons donc ce dataset :

# Nettoyage des données

In [14]:
print(f"Il y a {df_raw.select('CustomerID').distinct().count()} clients.")

Il y a 4373 clients.


In [40]:
df_cleanV1 = (
    df_raw
    .filter(col("CustomerID").isNotNull())
    #.filter(col("Quantity") > 0) 
    #.filter(col("UnitPrice") > 0) 
)

In [41]:
print(f"Il y a {df_cleanV1.select('CustomerID').distinct().count()} clients.")

Il y a 4372 clients.


La colonne InvoiceDate est initialement stockée sous forme de chaîne de caractères.  
Afin de permettre les calculs temporels nécessaires à l’analyse RFM, cette variable est convertie en format timestamp.

In [42]:
from pyspark.sql.functions import to_timestamp, col

df_cleanV1 = df_cleanV1.withColumn("InvoiceDate", to_timestamp(col("InvoiceDate"), "dd/MM/yyyy HH:mm"))
df_cleanV1.printSchema()
df_cleanV1.select("InvoiceDate").show(5)

root
 |-- InvoiceNo: string (nullable = true)
 |-- StockCode: string (nullable = true)
 |-- Description: string (nullable = true)
 |-- Quantity: integer (nullable = true)
 |-- InvoiceDate: timestamp (nullable = true)
 |-- UnitPrice: string (nullable = true)
 |-- CustomerID: integer (nullable = true)
 |-- Country: string (nullable = true)

+-------------------+
|        InvoiceDate|
+-------------------+
|2010-12-01 08:26:00|
|2010-12-01 08:26:00|
|2010-12-01 08:26:00|
|2010-12-01 08:26:00|
|2010-12-01 08:26:00|
+-------------------+
only showing top 5 rows



On transforme la variable UnitPrice en format double pour pouvoir creer une nouvelle colonne par la suite : 

In [43]:
from pyspark.sql.functions import regexp_replace

df_cleanV1 = df_cleanV1.withColumn("UnitPrice", regexp_replace(col("UnitPrice"), ",", "." ).cast("double"))

df_cleanV1.printSchema()

root
 |-- InvoiceNo: string (nullable = true)
 |-- StockCode: string (nullable = true)
 |-- Description: string (nullable = true)
 |-- Quantity: integer (nullable = true)
 |-- InvoiceDate: timestamp (nullable = true)
 |-- UnitPrice: double (nullable = true)
 |-- CustomerID: integer (nullable = true)
 |-- Country: string (nullable = true)



## Ajout de Colonnes

Création Colonne TotalAmount

In [44]:
df_cleanV1 = df_cleanV1.withColumn("TotalAmount", col("Quantity") * col("UnitPrice"))

print("Nouvelle colonne : TotalAmount = Quantity × UnitPrice")
df_cleanV1.show(10)

Nouvelle colonne : TotalAmount = Quantity × UnitPrice
+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+------------------+
|InvoiceNo|StockCode|         Description|Quantity|        InvoiceDate|UnitPrice|CustomerID|       Country|       TotalAmount|
+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+------------------+
|   536365|   85123A|WHITE HANGING HEA...|       6|2010-12-01 08:26:00|     2.55|     17850|United Kingdom|15.299999999999999|
|   536365|    71053| WHITE METAL LANTERN|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingdom|             20.34|
|   536365|   84406B|CREAM CUPID HEART...|       8|2010-12-01 08:26:00|     2.75|     17850|United Kingdom|              22.0|
|   536365|   84029G|KNITTED UNION FLA...|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingdom|             20.34|
|   536365|   84029E|RED WOOLLY HOTTIE...|       6|2010-1

Statistiques descriptives du dataset nettoyé

In [20]:
df_cleanV1.select("Quantity", "UnitPrice", "TotalAmount").describe().show()

+-------+------------------+-----------------+------------------+
|summary|          Quantity|        UnitPrice|       TotalAmount|
+-------+------------------+-----------------+------------------+
|  count|              1712|             1712|              1712|
|   mean| 5.136682242990654|35.89778037383178| 75.98072429906541|
| stddev|18.870440432713025|76.74851762139058|105.17974474580936|
|    min|                 1|              1.0|               1.0|
|    max|               392|           2500.0|            2500.0|
+-------+------------------+-----------------+------------------+



## Colonnes Type de transaction et NetAmount

In [45]:
#from pyspark.sql.functions import col, when, sum as spark_sum, countDistinct, datediff, lit, max as spark_max

# Ajouter la colonne NetAmount et TypeTransaction
#df_cleanV1 = df_cleanV1.withColumn(
#    "TypeTransaction",
#    when(col("TotalAmount") >= 0, "Achat").otherwise("Remboursement")
#).withColumn(
#    "NetAmount",
#    when(col("TypeTransaction") == "Achat", col("TotalAmount"))
#    .otherwise(col("TotalAmount"))  # remboursements sont soustraits
#)

#df_cleanV1.select("InvoiceNo", "CustomerID", "TotalAmount", "TypeTransaction", "NetAmount").show(10)


## Colonnes temporelles


Pour identifier les patterns d'achat (clients qui achètent plutôt le weekend, en soirée, etc.)

In [26]:
# Extraire les composantes de la date
#df_cleanV1 = df_cleanV1.withColumn("Year", year("InvoiceDate"))
#df_cleanV1 = df_cleanV1.withColumn("Month", month("InvoiceDate"))
#df_cleanV1 = df_cleanV1.withColumn("Quarter", quarter("InvoiceDate"))  # Trimestre (1-4)
#df_cleanV1 = df_cleanV1.withColumn("DayOfWeek", dayofweek("InvoiceDate"))  # 1=Dimanche, 7=Samedi
#df_cleanV1 = df_cleanV1.withColumn("Hour", hour("InvoiceDate"))

# Weekend ou semaine ?
#df_cleanV1 = df_cleanV1.withColumn(
#    "IsWeekend",
#    when((col("DayOfWeek") == 1) | (col("DayOfWeek") == 7), 1).otherwise(0)
#)

# Période de la journée
#df_cleanV1 = df_cleanV1.withColumn(
#    "TimeOfDay",
#    when((col("Hour") >= 6) & (col("Hour") < 12), "Morning")
#    .when((col("Hour") >= 12) & (col("Hour") < 18), "Afternoon")
#    .when((col("Hour") >= 18) & (col("Hour") < 22), "Evening")
#    .otherwise("Night")
#)

In [27]:
# Analyses colonnes temporelles
#print("\n--- Par période de la journée")
#df_cleanV1.groupBy("TimeOfDay").agg(
#    count("*").alias("Nb"),
#    spark_round(spark_sum("TotalAmount"), 0).alias("Revenu"),
 #   spark_round(spark_avg("TotalAmount"), 2).alias("Moyen")
#).show()

#print("\n--- Weekend vs Semaine ---")
#df_cleanV1.groupBy(when(col("IsWeekend") == 1, "Weekend").otherwise("Semaine").alias("Type")).agg(
 #   count("*").alias("Nb"),
  #  spark_round(spark_sum("TotalAmount"), 0).alias("Revenu"),
   # spark_round(spark_avg("TotalAmount"), 2).alias("Moyenne")
#).show()

#print("\n--- Par Mois ---")
#df_cleanV1.groupBy("Month").agg(
#    spark_round(spark_sum("TotalAmount"), 0).alias("Revenu"),
 #   spark_round(spark_avg("TotalAmount"), 2).alias("Moyenne")
#).orderBy("Month").show()

# 3) SEGMENTATION CLIENT (NON SUPERVISÉ)

 --- 1. CRÉATION DES VARIABLES RFM ---

Calculer la date de référence (date maximale dans le dataset)

In [34]:
df_cleanV1.describe().show()

+-------+------------------+------------------+--------------------+------------------+-----------------+------------------+--------------+------------------+---------------+------------------+
|summary|         InvoiceNo|         StockCode|         Description|          Quantity|        UnitPrice|        CustomerID|       Country|       TotalAmount|TypeTransaction|         NetAmount|
+-------+------------------+------------------+--------------------+------------------+-----------------+------------------+--------------+------------------+---------------+------------------+
|  count|              1712|              1712|                1712|              1712|             1712|              1712|          1712|              1712|           1712|              1712|
|   mean| 561188.4871495328|26069.022959183672|                NULL| 5.136682242990654|35.89778037383178|13595.035046728972|          NULL| 75.98072429906541|           NULL| 75.98072429906541|
| stddev|13102.849692316742| 1

In [25]:
reference_date = df_cleanV1.select(spark_max("InvoiceDate")).first()[0]
print(f"Date de référence pour Recency : {reference_date}")

Date de référence pour Recency : 2011-12-09 12:16:00


In [48]:
from pyspark.sql.functions import col, sum as spark_sum, countDistinct, max as spark_max, datediff, lit, when

rfm = df_cleanV1.groupBy("CustomerID").agg(
    spark_sum(col("TotalAmount")).alias("Monetary"),
    countDistinct(when(col("Quantity") > 0, col("InvoiceNo"))).alias("Frequency"),
    datediff(lit(reference_date), spark_max(when(col("Quantity") > 0, col("InvoiceDate")))).alias("Recency")
)

rfm_clean = rfm.filter(col("Monetary") > 0)
rfm_clean.show(10)

+----------+------------------+---------+-------+
|CustomerID|          Monetary|Frequency|Recency|
+----------+------------------+---------+-------+
|     15727|           5178.96|        7|     16|
|     17389|          31300.08|       34|      0|
|     16503|1421.4299999999998|        4|    106|
|     17420| 598.8299999999999|        3|     50|
|     15447|            155.17|        1|    330|
|     14450|            483.25|        3|    180|
|     13285|2709.1199999999994|        4|     23|
|     14570|218.05999999999997|        2|    280|
|     13623|            672.44|        5|     30|
|     16339|109.95000000000002|        1|    284|
+----------+------------------+---------+-------+
only showing top 10 rows



Le montant Monetary correspond au chiffre d’affaires net par client, intégrant les remboursements (montants négatifs).

In [52]:
rfm_clean.select("CustomerID").count()

4321

Calculer les variables RFM par client

In [53]:
# Calcul RFM par client
#rfm_data = df_cleanV1.groupBy("CustomerID").agg(
#    datediff(lit(reference_date), spark_max(col("InvoiceDate"))).alias("Recency"),
#    countDistinct("InvoiceNo").alias("Frequency"),
#    spark_sum("NetAmount").alias("Monetary")  # Utiliser NetAmount ici 
#)

# Vérifier le résultat
#rfm_data.show(10)
#rfm_data.describe().show()

### 2. Verification du Dataframe RFM

In [57]:
#rfm_clean.select([count(when(col(c).isNull(), 1)).alias(c) for c in rfm_clean.columns]).show()
rfm_clean.select("Recency", "Frequency", "Monetary").describe().show()


+-------+-----------------+-----------------+--------------------+
|summary|          Recency|        Frequency|            Monetary|
+-------+-----------------+-----------------+--------------------+
|  count|             4319|             4319|                4319|
|   mean| 91.7895346144941|4.286408890946978|   1924.819391988883|
| stddev|99.83713194497682|7.720342290736043|   8265.841921253997|
|    min|                0|                1|5.684341886080801...|
|    max|              373|              210|   279489.0199999999|
+-------+-----------------+-----------------+--------------------+



In [58]:
rfm_clean.approxQuantile(["Monetary", "Frequency"], [0.25, 0.5, 0.75, 0.95, 0.99], 0.05)

[[313.61999999999995,
  656.2500000000001,
  1415.29,
  279489.0199999999,
  279489.0199999999],
 [1.0, 2.0, 4.0, 210.0, 210.0]]

In [59]:
rfm_clean.orderBy(col("Monetary").desc()).show(10)

+----------+------------------+---------+-------+
|CustomerID|          Monetary|Frequency|Recency|
+----------+------------------+---------+-------+
|     14646| 279489.0199999999|       74|      1|
|     18102|256438.49000000005|       60|      0|
|     17450|187482.16999999993|       46|      8|
|     14911| 132572.6200000001|      201|      1|
|     12415|123725.45000000003|       21|     24|
|     14156|         113384.14|       55|      9|
|     17511| 88125.37999999999|       31|      2|
|     16684|          65892.08|       28|      4|
|     13694| 62653.09999999999|       50|      3|
|     15311|          59419.34|       91|      0|
+----------+------------------+---------+-------+
only showing top 10 rows



### 3. ASSEMBLAGE ET STANDARDISATION DES FEATURES

In [60]:
from pyspark.ml.feature import VectorAssembler, StandardScaler


In [61]:
# Assemblage des variables RFM
assembler = VectorAssembler(
    inputCols=["Recency", "Frequency", "Monetary"], 
    outputCol="features_raw"
)
rfm_assembled = assembler.transform(rfm_clean)

# Standardisation pour éviter que Monetary domine le clustering
scaler = StandardScaler(
    inputCol="features_raw", 
    outputCol="features", 
    withMean=True, 
    withStd=True
)
scaler_model = scaler.fit(rfm_assembled)
rfm_scaled = scaler_model.transform(rfm_assembled)

rfm_scaled.select("CustomerID", "Recency", "Frequency", "Monetary", "features").show(5, truncate=False)


+----------+-------+---------+------------------+---------------------------------------------------------------+
|CustomerID|Recency|Frequency|Monetary          |features                                                       |
+----------+-------+---------+------------------+---------------------------------------------------------------+
|15727     |16     |7        |5178.96           |[-0.7591317292273967,0.35148585475402605,0.39368531832719045]  |
|17389     |0      |34       |31300.08          |[-0.919392743223853,3.8487401192959507,3.5538134999265316]     |
|16503     |106    |4        |1421.4299999999998|[0.14233647450267023,-0.03709795241729891,-0.06089995390481845]|
|17420     |50     |3        |598.8299999999999 |[-0.418577074484927,-0.1666258881410739,-0.16041794709131343]  |
|15447     |330    |1        |155.17            |[2.385990670453059,-0.4256817595886239,-0.21409185039440162]   |
+----------+-------+---------+------------------+---------------------------------------

Les variables RFM (Recency, Frequency, Monetary) ont été assemblées à l’aide d’un VectorAssembler, puis standardisées via un StandardScaler afin d’éviter que la variable Monetary, de plus grande amplitude, ne domine le processus de clustering. Les données ainsi normalisées sont utilisées comme entrée du modèle de segmentation non supervisée.

### 4. Entrainement

Pour ce projet nous utilisons un clustering hiérarchique descendant scalable via BisectingKMeans. Ce choix a été pousser par les points suivants :  
- Très rapide
- Plus stable que KMeans classique
- Meilleure séparation quand les clusters sont déséquilibrés

In [62]:
rfm_scaled.select("CustomerID", "features").show(5, truncate=False)

+----------+---------------------------------------------------------------+
|CustomerID|features                                                       |
+----------+---------------------------------------------------------------+
|15727     |[-0.7591317292273967,0.35148585475402605,0.39368531832719045]  |
|17389     |[-0.919392743223853,3.8487401192959507,3.5538134999265316]     |
|16503     |[0.14233647450267023,-0.03709795241729891,-0.06089995390481845]|
|17420     |[-0.418577074484927,-0.1666258881410739,-0.16041794709131343]  |
|15447     |[2.385990670453059,-0.4256817595886239,-0.21409185039440162]   |
+----------+---------------------------------------------------------------+
only showing top 5 rows



Pour déterminer le nombre de clusters, nous utiliserons le silhouette score. En effet, nous n'utiliserons pas la méthode du coude car en Spark, le coude est moins fiable, et pas toujours clair visuellement.

In [63]:
from pyspark.ml.clustering import BisectingKMeans
from pyspark.ml.evaluation import ClusteringEvaluator

In [70]:
from pyspark.sql.functions import countDistinct

silhouette_scores = []

for k in range(2, 9):
    bkm = BisectingKMeans(
        k=k,
        featuresCol="features",
        predictionCol="cluster",
        seed=42
    )

    model = bkm.fit(rfm_scaled)
    predictions = model.transform(rfm_scaled)

    nb_clusters = predictions.select("cluster").distinct().count()

    if nb_clusters > 1:
        score = evaluator.evaluate(predictions)
        silhouette_scores.append((k, score))
        print(f"k = {k}, silhouette = {score:.4f}")
    else:
        print(f"k = {k}, 1 seul cluster : silhouette non calculable")



k = 2, 1 seul cluster : silhouette non calculable
k = 3, 1 seul cluster : silhouette non calculable
k = 4, 1 seul cluster : silhouette non calculable
k = 5, 1 seul cluster : silhouette non calculable
k = 6, 1 seul cluster : silhouette non calculable
k = 7, 1 seul cluster : silhouette non calculable
k = 8, 1 seul cluster : silhouette non calculable


BisectingKMeans n’a pas permis de produire plusieurs clusters sur les variables RFM, ce qui indique une structure compacte des données. Nous avons donc retenu un autre algorithme plus adapté.